File: 027_Two Sum II - Input Array Is Sorted.py
"""
Two Sum II - Input Array Is Sorted

Find two numbers (1-indexed) adding to target in sorted array.
"""

from typing import List, Tuple

def two_sum(numbers: List[int], target: int) -> Tuple[int,int]:
    l, r = 0, len(numbers) - 1
    while l < r:
        s = numbers[l] + numbers[r]
        if s == target:
            return (l+1, r+1)
        if s < target:
            l += 1
        else:
            r -= 1
    return (-1,-1)


if __name__ == '__main__':
    print('TwoSumII 1:', two_sum([2,7,11,15], 9))
    print('TwoSumII 2:', two_sum([2,3,4], 6))

# Time Complexity: O(n)
# Space Complexity: O(1)

File: 027_Two Sum II - Input Array Is Sorted.cpp
/*
Two Sum II - Input Array Is Sorted

Given a 1-indexed sorted array of integers, find two numbers such that they add up to a specific target. Return indices (1-indexed).

Strategy: Two-pointer from both ends.
*/

#include <vector>
#include <iostream>
using namespace std;

pair<int,int> twoSum(vector<int>& numbers, int target) {
    int l = 0, r = numbers.size() - 1;
    while (l < r) {
        int sum = numbers[l] + numbers[r];
        if (sum == target) return {l+1, r+1};
        if (sum < target) ++l; else --r;
    }
    return {-1,-1};
}

int main() {
    vector<int> a1 = {2,7,11,15};
    auto p = twoSum(a1, 9);
    cout << "TwoSumII 1: " << p.first << ',' << p.second << '\n';
    vector<int> a2 = {2,3,4};
    auto p2 = twoSum(a2, 6);
    cout << "TwoSumII 2: " << p2.first << ',' << p2.second << '\n';
    return 0;
}

/*
Time Complexity: O(n)
Space Complexity: O(1)
*/
